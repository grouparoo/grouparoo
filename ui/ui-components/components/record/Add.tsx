import { useState } from "react";
import { UseApi } from "../../hooks/useApi";
import { useRouter } from "next/router";
import LoadingButton from "../LoadingButton";
import { Actions } from "../../utils/apiData";

export default function AddRecord(props) {
  const { execApi } = UseApi(props);
  const router = useRouter();
  const [loading, setLoading] = useState(false);

  async function create() {
    setLoading(true);

    const { record }: Actions.RecordCreate = await execApi("post", `/record`);
    router.push(
      "/model/[modelId]/record/[recordId]/edit",
      `/model/${record.modelId}/record/${record.id}/edit`
    );
  }

  return (
    <>
      <LoadingButton variant="warning" disabled={loading} onClick={create}>
        Add new Record
      </LoadingButton>
    </>
  );
}
