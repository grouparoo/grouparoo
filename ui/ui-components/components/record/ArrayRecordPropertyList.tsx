import { Button } from "react-bootstrap";
import { Fragment, useState } from "react";
import { formatTimestamp } from "../../utils/formatTimestamp";

const limit = 5;

export default function ArrayRecordPropertyList({
  type,
  values,
  invalidValue,
  invalidReason,
}: {
  type: string;
  values: (string | number | boolean | Date)[];
  invalidValue: string;
  invalidReason: string;
}) {
  const [expanded, setExpanded] = useState(false);
  if (!values || values.length === 0) return null;

  let formattedValues = values.map((value) => {
    if (value === null) {
      return <code>null</code>;
    } else if (value === true || value === false) {
      // return <input type="checkbox" checked={value} readOnly />;
      return <>{value.toString()}</>;
    } else if (type === "date") {
      return <>{formatTimestamp(value)}</>;
    } else {
      return <>{value}</>;
    }
  });

  return (
    <>
      {formattedValues
        .filter((V, idx) => (idx < limit || expanded ? true : false))
        .map((V, idx) => (
          <Fragment key={`id=${idx}`}>
            {V}
            <br />
          </Fragment>
        ))}

      {invalidValue ? (
        <span className="text-danger">
          invalid value: {invalidValue} | {invalidReason}
        </span>
      ) : null}

      {formattedValues.length > limit && !expanded ? (
        <Button
          size="sm"
          variant="clear"
          onClick={() => setExpanded(!expanded)}
        >
          <u>... more</u>
        </Button>
      ) : null}
    </>
  );
}
