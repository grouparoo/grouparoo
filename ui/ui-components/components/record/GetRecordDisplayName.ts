import { Models } from "../../utils/apiData";

export function getRecordDisplayName(record: Models.GrouparooRecordType) {
  let displayName = `Record: ${record.id}`;

  for (const [key, property] of Object.entries(record.properties)) {
    if (property.isPrimaryKey) {
      displayName = `${key}: ${property.values.join(", ")}`;
      break;
    }
  }

  return displayName;
}
