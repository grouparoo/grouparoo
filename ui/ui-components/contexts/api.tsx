import { createContext, useContext } from "react";
import { Client } from "../client/client";

export const ApiContext = createContext({
  client: new Client(),
});
export const useApi = () => useContext(ApiContext);
