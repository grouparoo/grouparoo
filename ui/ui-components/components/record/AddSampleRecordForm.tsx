import { useCallback, useMemo, useState } from "react";
import { Form } from "react-bootstrap";
import { useForm } from "react-hook-form";
import { useApi } from "../../contexts/api";
import { errorHandler } from "../../eventHandlers";
import { Actions, Models } from "../../utils/apiData";
import LoadingButton from "../LoadingButton";

const getInputType = (type?: Models.PropertyType["type"]): string => {
  switch (type) {
    case "float":
    case "integer":
      return "number";
    default:
      return "text";
  }
};

interface Props {
  modelId: string;
  properties: Models.PropertyType[];
  onSubmitComplete: (record?: Models.GrouparooRecordType) => void;
}

const AddSampleRecordForm: React.FC<Props> = ({
  modelId,
  onSubmitComplete,
  properties,
}) => {
  const { handleSubmit, register } = useForm();
  const [submitting, setSubmitting] = useState(false);
  const { client } = useApi();

  const propertiesWithPrimaryKey = useMemo(
    () =>
      properties.filter((property) => property.unique && property.isPrimaryKey),
    [properties]
  );

  const [selectedUniquePropertyValue, setSelectedUniquePropertyValue] =
    useState(propertiesWithPrimaryKey?.[0]?.key || "");

  const selectedUniqueProperty = useMemo(() => {
    return propertiesWithPrimaryKey?.find(
      ({ key }) => key === selectedUniquePropertyValue
    );
  }, [propertiesWithPrimaryKey, selectedUniquePropertyValue]);

  const onSelectUniqueProperty: React.ChangeEventHandler<HTMLInputElement> = (
    event
  ) => {
    setSelectedUniquePropertyValue(event.target.value);
  };

  const onSubmit: Parameters<typeof handleSubmit>[0] = useCallback(
    async (data) => {
      setSubmitting(true);
      const response = await client.request<Actions.RecordCreate>(
        "post",
        `/record`,
        {
          modelId,
          properties: { [data.uniqueProperty]: data.value },
        }
      );
      setSubmitting(false);

      if (response?.record) {
        onSubmitComplete(response.record);
      } else {
        errorHandler.set({
          message: `Could not add Sample Record.
          Record with ${data.uniqueProperty} = ${data.value} may not exist or may be already added.`,
        });
      }
    },
    [client, modelId, onSubmitComplete]
  );

  return (
    <Form id="form" onSubmit={handleSubmit(onSubmit)}>
      <Form.Group>
        <Form.Label>Unique Property</Form.Label>
        <Form.Control
          name="uniqueProperty"
          as="select"
          ref={register}
          disabled={submitting}
          value={selectedUniquePropertyValue}
          onChange={onSelectUniqueProperty}
        >
          {propertiesWithPrimaryKey.map((rule) => (
            <option key={`rule-${rule.key}`} value={rule.key}>
              {rule.key}
            </option>
          ))}
        </Form.Control>
      </Form.Group>
      <Form.Group>
        <Form.Label>Value</Form.Label>
        <Form.Control
          autoFocus
          required
          type={getInputType(selectedUniqueProperty?.type)}
          name="value"
          ref={register}
          disabled={submitting}
        />
        <Form.Control.Feedback type="invalid">
          Value is required
        </Form.Control.Feedback>
      </Form.Group>

      <LoadingButton variant="primary" type="submit" disabled={submitting}>
        Submit
      </LoadingButton>
    </Form>
  );
};

export default AddSampleRecordForm;
