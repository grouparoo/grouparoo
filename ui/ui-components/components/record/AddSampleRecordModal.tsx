import { Modal } from "react-bootstrap";
import { Models } from "../../utils/apiData";
import AddSampleRecordForm from "./AddSampleRecordForm";
import type { RecordType } from "./SampleRecordCard";

interface Props {
  show: boolean;
  onRecordCreated: (recordId: string) => void;
  onHide: () => void;
  properties: Models.PropertyType[];
}

const AddSampleRecordModal: React.FC<Props> = ({
  properties,
  show,
  onHide,
  onRecordCreated,
}) => {
  return (
    <Modal show={show} onHide={onHide} centered>
      <Modal.Header closeButton>
        <Modal.Title as="h6">Add Sample Record</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        <AddSampleRecordForm
          properties={properties}
          onSubmitComplete={(recordId) => {
            onRecordCreated(recordId);
            onHide();
          }}
        />
      </Modal.Body>
    </Modal>
  );
};

export default AddSampleRecordModal;
