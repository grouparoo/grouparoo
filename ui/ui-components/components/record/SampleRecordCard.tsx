import Link from "next/link";
import React, {
  Fragment,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";
import { Card, Col, Row, Table } from "react-bootstrap";
import EnterpriseLink from "../../components/GrouparooLink";
import { ApiHook } from "../../hooks/useApi";
import { usePrevious } from "../../hooks/usePrevious";
import { Actions, Models } from "../../utils/apiData";
import { errorHandler, successHandler } from "../../eventHandlers";
import { grouparooUiEdition } from "../../utils/uiEdition";
import StateBadge from "../badges/StateBadge";
import ManagedCard from "../lib/ManagedCard";
import SeparatedItems from "../lib/SeparatedItems";
import LinkButton from "../LinkButton";
import Loader from "../Loader";
import LoadingButton from "../LoadingButton";
import RecordImageFromEmail from "../visualizations/RecordImageFromEmail";
import AddSampleRecordModal from "./AddSampleRecordModal";
import ArrayRecordPropertyList from "./ArrayRecordPropertyList";

export type RecordType =
  | Models.GrouparooRecordType
  | Models.DestinationRecordPreviewType;

export interface SampleRecordCardProps {
  modelId: string;
  execApi: ApiHook["execApi"];
  fetchRecord: (recordId?: string) => Promise<{
    record?: RecordType;
    groups?: Models.GroupType[];
    destinations?: Models.DestinationType[];
  }>;
  allowFetchWithoutRecordId?: boolean;
  properties: Models.PropertyType[];
  propertiesTitle?: string;
  groupsTitle?: string;
  disabled?: boolean;
  hideViewAllRecords?: boolean;
  highlightProperty?: Models.PropertyType;
  highlightPropertyError?: string;
  propertyLinkDisabled?: boolean;
  importDisabled?: boolean;
  reloadKey?: string;
  warning?: string;
}

// This is exported so we can manipulate it in tests, but is unused elsewhere.
export const isConfigUI = grouparooUiEdition() === "config";

const getCachedSampleRecordId = (modelId: string): string => {
  return globalThis.localStorage?.getItem(
    `sampleRecord:${grouparooUiEdition()}:${modelId}`
  );
};

const setCachedSampleRecordId = (modelId: string, recordId: string): void => {
  if (recordId) {
    globalThis.localStorage?.setItem(
      `sampleRecord:${grouparooUiEdition()}:${modelId}`,
      recordId
    );
  }
};

const clearCachedSampleRecordId = (modelId: string): void => {
  globalThis.localStorage?.removeItem(
    `sampleRecord:${grouparooUiEdition()}:${modelId}`
  );
};

const SampleRecordCard: React.FC<SampleRecordCardProps> = ({
  modelId,
  properties,
  propertiesTitle,
  execApi,
  fetchRecord,
  groupsTitle = "Groups",
  disabled = false,
  importDisabled = false,
  hideViewAllRecords = false,
  highlightProperty,
  highlightPropertyError,
  propertyLinkDisabled = false,
  allowFetchWithoutRecordId = false,
  warning,
  reloadKey,
}) => {
  const prevModelId = usePrevious(modelId);
  const prevReloadKey = usePrevious(reloadKey);

  const [loading, setLoading] = useState(false);
  const [importing, setImporting] = useState(false);
  const [addingRecord, setAddingRecord] = useState(false);
  const [hasRecords, setHasRecords] = useState(true);
  const [record, setRecord] = useState<RecordType>();
  const [groups, setGroups] = useState<Models.GroupType[]>();
  const [destinations, setDestinations] = useState<Models.DestinationType[]>();
  const [recordId, setRecordId] = useState(() =>
    getCachedSampleRecordId(modelId)
  );

  const saveRecord = useCallback(
    (record, groups = undefined, destinations = undefined) => {
      setLoading(true);
      setRecord(record);
      setGroups(groups);
      setDestinations(destinations);
      setRecordId(record.id);
      setHasRecords(true);
      setLoading(false);

      setCachedSampleRecordId(modelId, record.id);
    },
    [modelId]
  );

  const loadRecord = useCallback(async () => {
    setLoading(true);

    let record: Models.GrouparooRecordType;
    let groups: Models.GroupType[];
    let destinations: Models.DestinationType[];

    if (recordId || allowFetchWithoutRecordId) {
      ({ record, groups, destinations } = await fetchRecord(recordId));

      if (allowFetchWithoutRecordId && !record && !recordId) {
        setHasRecords(false);
        errorHandler.set({
          message:
            "Could not switch Sample Record because no Records were found.",
        });
        setRecordId(getCachedSampleRecordId(modelId));
        setLoading(false);
        return;
      }
    } else {
      ({ record, groups, destinations } = await execApi<Actions.RecordsList>(
        "get",
        "/records",
        {
          limit: 25,
          offset: 0,
          modelId,
        }
      ).then((response) => {
        if (response.total === 0) {
          setHasRecords(false);
        } else {
          const randomRecord = response?.records?.length
            ? response.records[
                Math.floor(Math.random() * response.records.length)
              ]
            : undefined;

          if (randomRecord?.id) {
            return fetchRecord(randomRecord?.id);
          }
        }

        return {
          record: undefined,
          groups: undefined,
          destinations: undefined,
        };
      }));
    }

    if (record?.id) {
      saveRecord(record, groups, destinations);
    } else {
      // Got an invalid record. Let's clear this and start over.
      clearCachedSampleRecordId(modelId);
      setRecordId(undefined);
    }

    setLoading(false);
  }, [
    recordId,
    allowFetchWithoutRecordId,
    fetchRecord,
    modelId,
    execApi,
    saveRecord,
  ]);

  const sortedPropertyKeys = useMemo(() => {
    const id = highlightProperty?.id;
    return record?.properties
      ? Object.keys(record.properties).sort((a, b) =>
          id === record.properties[a].id
            ? -1
            : id === record.properties[b].id
            ? 1
            : a.localeCompare(b)
        )
      : undefined;
  }, [highlightProperty?.id, record]);

  useEffect(() => {
    if (prevReloadKey !== reloadKey) {
      setHasRecords(true);
    }
  }, [prevReloadKey, reloadKey, recordId]);

  useEffect(() => {
    // Switched to another model
    if (prevModelId && prevModelId !== modelId) {
      setLoading(true);
      setRecordId(getCachedSampleRecordId(modelId));
      setRecord(undefined);
      setGroups(undefined);
      setDestinations(undefined);
      setHasRecords(true);
      setLoading(false);
    }
  }, [record, modelId, prevModelId]);

  useEffect(() => {
    if (
      !loading &&
      hasRecords &&
      (!record || record.id !== recordId || reloadKey !== prevReloadKey)
    ) {
      loadRecord();
    }
  }, [
    record,
    recordId,
    loading,
    loadRecord,
    hasRecords,
    reloadKey,
    prevReloadKey,
  ]);

  const email = useMemo<string>(() => {
    if (!record) return undefined;

    const emailProperty = Object.values(record.properties).find(
      (property) => property.type === "email" && property.values.length > 0
    );

    return emailProperty?.values?.[0]?.toString() ?? undefined;
  }, [record]);

  const cardActions = useMemo(() => {
    const result: React.ReactNode[] = [];
    if (record) {
      result.push(
        <LinkButton
          key={modelId + (record?.id || "")}
          disabled={!record || disabled}
          href={`/model/${modelId}/record${
            record ? `/${record.id}/edit` : "s"
          }`}
          size="sm"
          variant="outline-primary"
        >
          View Record
        </LinkButton>
      );
    }
    if (!hideViewAllRecords && hasRecords) {
      result.push(
        <LinkButton
          disabled={disabled}
          href={`/model/${modelId}/records`}
          size="sm"
          variant="outline-primary"
        >
          View all Records
        </LinkButton>
      );
    }
    return result;
  }, [record, hideViewAllRecords, hasRecords, modelId, disabled]);

  if (!sortedPropertyKeys?.length && !properties.length) {
    return (
      <ManagedCard title="Sample Record">
        <Card.Body>
          {loading ? (
            <div className="text-center">
              <Loader size="sm" />
            </div>
          ) : (
            "At least one Grouparoo Property needs to be mapped before showing the Sample Record."
          )}
        </Card.Body>
      </ManagedCard>
    );
  }

  const importRecord = async () => {
    setImporting(true);
    const response = await execApi<Actions.RecordImport>(
      "post",
      `/record/${record.id}/import`
    );
    if (response?.record) {
      setRecord(response.record);
      setGroups(response.groups);
      successHandler.set({ message: "Import Complete!" });
    } else {
      loadRecord(); // we may have done a partial import
    }
    setImporting(false);
  };

  const actions = [
    <LoadingButton
      key="clear-record-id"
      disabled={disabled || !hasRecords || loading || importing}
      loading={loading && !recordId}
      size="sm"
      variant="outline-primary"
      onClick={() => {
        setRecordId(undefined);
      }}
    >
      Switch to random Record
    </LoadingButton>,
  ];

  if (!importDisabled) {
    actions.unshift(
      <LoadingButton
        disabled={disabled || !record || loading}
        loading={importing}
        onClick={() => {
          importRecord();
        }}
        size="sm"
        variant="outline-success"
      >
        Import Record Data
      </LoadingButton>
    );
  }

  if (isConfigUI) {
    actions.push(
      <LoadingButton
        disabled={disabled || loading}
        size="sm"
        variant="outline-primary"
        onClick={() => setAddingRecord(true)}
      >
        Add Sample Record
      </LoadingButton>
    );
  }

  const content = record ? (
    <Row>
      <Col style={{ minWidth: "75%" }}>
        {propertiesTitle && <h6>{propertiesTitle}</h6>}
        <Table bordered>
          <colgroup>
            <col span={1} style={{ width: "35%" }} />
            <col span={1} style={{ width: "65%" }} />
          </colgroup>
          <tbody>
            {sortedPropertyKeys &&
              sortedPropertyKeys.map((key, index) => {
                const property = record.properties[key];
                const isHighlightedProperty =
                  highlightProperty?.id === property.id;
                const trClassName = isHighlightedProperty
                  ? `table-${highlightPropertyError ? "danger" : "success"}`
                  : undefined;
                const name =
                  isHighlightedProperty && highlightProperty.key !== ""
                    ? highlightProperty.key
                    : key === ""
                    ? "Draft"
                    : key;

                return (
                  <tr
                    key={`property-${!key ? `draft-${index}` : key}`}
                    className={trClassName}
                  >
                    <th scope="row">
                      {isHighlightedProperty || propertyLinkDisabled ? (
                        name
                      ) : (
                        <EnterpriseLink
                          href={`/model/${modelId}/source/${property.sourceId}/property/${property.id}/edit`}
                        >
                          <a>{name}</a>
                        </EnterpriseLink>
                      )}
                      {property.state !== "ready" && (
                        <StateBadge state={property.state} marginBottom={0} />
                      )}
                    </th>
                    <td>
                      {isHighlightedProperty &&
                      (property.values.length === 0 ||
                        highlightPropertyError) ? (
                        `${
                          highlightPropertyError
                            ? highlightPropertyError
                            : "..."
                        }`
                      ) : (
                        <ArrayRecordPropertyList
                          type={property.type}
                          values={property.values}
                          invalidValue={property.invalidValue}
                          invalidReason={property.invalidReason}
                        />
                      )}
                    </td>
                  </tr>
                );
              })}
          </tbody>
        </Table>
      </Col>
      <Col
        className={"text-center"}
        xs={{ order: "first" }}
        md={{ order: "last" }}
      >
        {email && (
          <RecordImageFromEmail email={email} width={72} className="mb-4" />
        )}
        <h6>{groupsTitle}</h6>
        <p>
          {groups?.length
            ? groups.map((group) => {
                return (
                  <Fragment key={`group-${group.id}`}>
                    <EnterpriseLink
                      href={`/model/${group.modelId}/group/${group.id}/rules`}
                    >
                      {group.name}
                    </EnterpriseLink>
                    <br />
                  </Fragment>
                );
              })
            : (record as Models.DestinationRecordPreviewType)?.groupNames
                ?.length
            ? (record as Models.DestinationRecordPreviewType).groupNames.map(
                (groupName) => (
                  <Fragment key={`group-name-${groupName}`}>
                    {groupName}
                    <br />
                  </Fragment>
                )
              )
            : "None"}
        </p>
        {destinations && (
          <>
            <h6>Destinations</h6>
            <p>
              {destinations?.length
                ? destinations.map((destination) => {
                    return (
                      <Fragment key={`destination-${destination.id}`}>
                        {isConfigUI ? (
                          destination.name
                        ) : (
                          <EnterpriseLink
                            href={`/model/${destination.modelId}/destination/${destination.id}/data`}
                          >
                            <a>{destination.name}</a>
                          </EnterpriseLink>
                        )}
                        <br />
                      </Fragment>
                    );
                  })
                : "None"}
            </p>
          </>
        )}
      </Col>
    </Row>
  ) : (
    <Row>
      <Col>
        <p>
          A Sample Record can be used to validate your configuration is
          importing the correct data to Groups and Destinations.
        </p>
        {!isConfigUI && (
          <p>Run a Schedule on your Primary Source to generate Records.</p>
        )}
      </Col>
    </Row>
  );

  return (
    <ManagedCard
      title={<>Sample Record {loading && <Loader size="sm" />}</>}
      disabled={disabled}
      actions={cardActions}
      border={warning ? "warning" : undefined}
    >
      <Card.Body>
        {content}
        <Row>
          <Col>
            <SeparatedItems items={!hasRecords && !isConfigUI ? [] : actions} />
          </Col>
        </Row>
      </Card.Body>
      {isConfigUI && (
        <AddSampleRecordModal
          modelId={modelId}
          properties={properties}
          execApi={execApi}
          show={addingRecord}
          onRecordCreated={saveRecord}
          onHide={() => {
            setAddingRecord(false);
          }}
        />
      )}
      {warning && <Card.Footer>{warning}</Card.Footer>}
    </ManagedCard>
  );
};

export default SampleRecordCard;
