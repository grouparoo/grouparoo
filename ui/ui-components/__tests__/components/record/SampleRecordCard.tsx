import { act, render } from "@testing-library/react";
import SampleRecordCard, {
  RecordType,
  SampleRecordCardProps,
} from "../../../components/record/SampleRecordCard";
import { ApiContext } from "../../../contexts/api";
import { Models } from "../../../utils/apiData";

describe("SampleRecordCard", () => {
  let cardProps: SampleRecordCardProps;

  beforeEach(() => {
    cardProps = {
      modelId: "test-model",
      fetchRecord: async () => ({}),
      properties: [],
    };
    require("../../../components/record/SampleRecordCard").isConfigUI = false;
  });

  it("should render with no records", async () => {
    cardProps.properties = [{ id: "some-property", sourceId: "some-source" }];

    let card: ReturnType<typeof render>;
    await act(async () => {
      card = render(
        <ApiContext.Provider
          value={{
            client: {
              request: async () => ({
                records: [],
                total: 0,
              }),
            } as any,
          }}
        >
          <SampleRecordCard {...cardProps} />
        </ApiContext.Provider>
      );
    });
    expect(card.container).toMatchSnapshot();
  });

  it("should render with no records in 'config mode'", async () => {
    require("../../../components/record/SampleRecordCard").isConfigUI = true;
    cardProps.properties = [{ id: "some-property", sourceId: "some-source" }];

    let card: ReturnType<typeof render>;
    await act(async () => {
      card = render(
        <ApiContext.Provider
          value={{
            client: {
              request: async () => ({
                records: [],
                total: 0,
              }),
            } as any,
          }}
        >
          <SampleRecordCard {...cardProps} />
        </ApiContext.Provider>
      );
    });
    expect(card.container).toMatchSnapshot();
  });

  it("should render with 1 record", async () => {
    const record: RecordType = {
      id: "test-record",
      modelId: "test-model",
      properties: {
        asdf: {
          id: "fake-property",
          sourceId: "some-source",
          configId: "fake-config-id",
          state: "pending",
          values: [2],
          type: "integer",
          unique: false,
          isPrimaryKey: true,
          isArray: false,
          confirmedAt: new Date(1000),
          valueChangedAt: new Date(1000),
          updatedAt: new Date(1000),
          stateChangedAt: new Date(1000),
          startedAt: new Date(1000),
          invalidValue: null,
          invalidReason: null,
          createdAt: new Date(1000),
        },
      },
    };
    const fetchRecord: (recordId?: string) => Promise<{
      record?: RecordType;
      groups?: Models.GroupType[];
      destinations?: Models.DestinationType[];
    }> = async () => {
      return {
        record,
        groups: [],
        destinations: [],
      };
    };
    cardProps.fetchRecord = fetchRecord;

    let card: ReturnType<typeof render>;
    await act(async () => {
      card = render(
        <ApiContext.Provider
          value={{
            client: {
              request: async () => ({
                records: [record],
                total: 1,
              }),
            } as any,
          }}
        >
          <SampleRecordCard {...cardProps} />
        </ApiContext.Provider>
      );
    });
    expect(card.container).toMatchSnapshot();
  });

  it("should render with 1 record in config mode", async () => {
    require("../../../components/record/SampleRecordCard").isConfigUI = true;
    const record: RecordType = {
      id: "test-record",
      modelId: "test-model",
      properties: {
        asdf: {
          id: "fake-property",
          sourceId: "some-source",
          configId: "fake-config-id",
          state: "pending",
          values: [2],
          type: "integer",
          unique: false,
          isPrimaryKey: true,
          isArray: false,
          confirmedAt: new Date(1000),
          valueChangedAt: new Date(1000),
          updatedAt: new Date(1000),
          stateChangedAt: new Date(1000),
          startedAt: new Date(1000),
          invalidValue: null,
          invalidReason: null,
          createdAt: new Date(1000),
        },
      },
    };
    const fetchRecord: (recordId?: string) => Promise<{
      record?: RecordType;
      groups?: Models.GroupType[];
      destinations?: Models.DestinationType[];
    }> = async () => {
      return {
        record,
        groups: [],
        destinations: [],
      };
    };
    cardProps.fetchRecord = fetchRecord;

    let card: ReturnType<typeof render>;
    await act(async () => {
      card = render(
        <ApiContext.Provider
          value={{
            client: {
              request: async () => ({
                records: [record],
                total: 1,
              }),
            } as any,
          }}
        >
          <SampleRecordCard {...cardProps} />
        </ApiContext.Provider>
      );
    });
    expect(card.container).toMatchSnapshot();
  });
});
