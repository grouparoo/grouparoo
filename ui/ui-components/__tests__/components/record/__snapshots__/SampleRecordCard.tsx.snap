// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SampleRecordCard should render loading 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h5
            class="my-1"
          >
            Sample Record
          </h5>
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="text-center"
      >
        <div
          class="spinner-border spinner-border-sm"
          role="status"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SampleRecordCard should render with 1 record 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h5
            class="my-1"
          >
            Sample Record 
          </h5>
        </div>
        <div
          class="text-right col-lg-6 col-md-7 col-sm-12"
        >
          <a
            href="/model/test-model/record/test-record/edit"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View Record
            </button>
          </a>
            
          <a
            href="/model/test-model/records"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View All 
              1
               Records
            </button>
          </a>
          
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
          style="min-width: 75%;"
        >
          <table
            class="table table-bordered"
          >
            <colgroup>
              <col
                span="1"
                style="width: 35%;"
              />
              <col
                span="1"
                style="width: 65%;"
              />
            </colgroup>
            <tbody>
              <tr>
                <th
                  scope="row"
                >
                  <span>
                    <a>
                      asdf
                    </a>
                  </span>
                  <span
                    class="badge badge-warning"
                    style="margin-left: 3px; margin-right: 3px; margin-bottom: 0px;"
                  >
                    pending
                  </span>
                </th>
                <td>
                  2
                  <br />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="text-center col-md col order-md-last order-first"
        >
          <h6>
            Groups
          </h6>
          <p>
            None
          </p>
          <h6>
            Destinations
          </h6>
          <p>
            None
          </p>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <button
            class="btn btn-outline-success btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Import Record Data
            </span>
          </button>
            
          <button
            class="btn btn-outline-primary btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Switch to random Record
            </span>
          </button>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SampleRecordCard should render with 1 record in config mode 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h5
            class="my-1"
          >
            Sample Record 
          </h5>
        </div>
        <div
          class="text-right col-lg-6 col-md-7 col-sm-12"
        >
          <a
            href="/model/test-model/record/test-record/edit"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View Record
            </button>
          </a>
            
          <a
            href="/model/test-model/records"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View All 
              1
               Records
            </button>
          </a>
          
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
          style="min-width: 75%;"
        >
          <table
            class="table table-bordered"
          >
            <colgroup>
              <col
                span="1"
                style="width: 35%;"
              />
              <col
                span="1"
                style="width: 65%;"
              />
            </colgroup>
            <tbody>
              <tr>
                <th
                  scope="row"
                >
                  <span>
                    <a>
                      asdf
                    </a>
                  </span>
                  <span
                    class="badge badge-warning"
                    style="margin-left: 3px; margin-right: 3px; margin-bottom: 0px;"
                  >
                    pending
                  </span>
                </th>
                <td>
                  2
                  <br />
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="text-center col-md col order-md-last order-first"
        >
          <h6>
            Groups
          </h6>
          <p>
            None
          </p>
          <h6>
            Destinations
          </h6>
          <p>
            None
          </p>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <button
            class="btn btn-outline-success btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Import Record Data
            </span>
          </button>
            
          <button
            class="btn btn-outline-primary btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Switch to random Record
            </span>
          </button>
            
          <button
            class="btn btn-outline-primary btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Add Sample Record
            </span>
          </button>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SampleRecordCard should render with no records 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h5
            class="my-1"
          >
            Sample Record 
          </h5>
        </div>
        <div
          class="text-right col-lg-6 col-md-7 col-sm-12"
        >
          <a
            href="/model/test-model/records"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View All 
              0
               Records
            </button>
          </a>
          
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <p>
            A Sample Record can be used to validate your configuration is importing the correct data to Groups and Destinations.
          </p>
          <p>
            Run a Schedule on your Primary Source to generate Records.
          </p>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`SampleRecordCard should render with no records in 'config mode' 1`] = `
<div>
  <div
    class="card"
  >
    <div
      class="card-header"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <h5
            class="my-1"
          >
            Sample Record 
          </h5>
        </div>
        <div
          class="text-right col-lg-6 col-md-7 col-sm-12"
        >
          <a
            href="/model/test-model/records"
          >
            <button
              class="btn btn-outline-primary btn-sm"
              type="button"
            >
              View All 
              0
               Records
            </button>
          </a>
          
        </div>
      </div>
    </div>
    <div
      class="card-body"
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <p>
            A Sample Record can be used to validate your configuration is importing the correct data to Groups and Destinations.
          </p>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col"
        >
          <button
            class="btn btn-outline-success btn-sm"
            disabled=""
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Import Record Data
            </span>
          </button>
            
          <button
            class="btn btn-outline-primary btn-sm"
            disabled=""
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Switch to random Record
            </span>
          </button>
            
          <button
            class="btn btn-outline-primary btn-sm"
            style="display: inline-flex; align-items: center; justify-content: center;"
            type="button"
          >
            <span
              style="opacity: 1;"
            >
              Add Sample Record
            </span>
          </button>
          
        </div>
      </div>
    </div>
  </div>
</div>
`;
